### PR TITLE
finetune: Add -ngl parameter

### LINKED ISSUE #3458: finetune: -ngl option to offload to gpu?
How to you offload layers to the GPU with finetune? There is no `-ngl` option.

### PR DESCRIPTION
**Pull request by AndrewGodfrey**
Fix #3458

### CODE REVIEW

File: examples/finetune/finetune.sh [hunk 1]
```diff
@@ -0,0 +1,31 @@
+#!/bin/bash
+cd `dirname $0`
+cd ../..
+
+EXE="./finetune"
+
+# MODEL="c:/models/openllama-3b-v2-q8_0.gguf" # This is the model the readme uses.
+MODEL="c:/models/openllama-3b-v2.gguf" # An f16 model. Note in this case with "-g", you get an f32-format .BIN file that isn't yet supported if you use it with "main --lora" with GPU inferencing.
```
Comment (cebtenzzre) [inline examples/finetune/finetune.sh:8]: Hardcoding an absolute Windows path inside of a _bash_ script (Unix-like) is an assumption that I think will not apply to many users. Better to use a relative path.

Same issue with --train-data below.

**Comment by ggerganov** [comment]
The `train` and `finetune` examples do not support GPU offloading. There is no point in adding `-ngl` parameter

**Comment by AndrewGodfrey** [comment]
> The `train` and `finetune` examples do not support GPU offloading. There is no point in adding `-ngl` parameter

I believe it did work for me (with these changes) for f32, with an f16 base model. The result was something I couldn't directly use with --lora and GPU inference, but I *could* use it with --lora and CPU inference, and it did work (I verified that it had finetuned). The only way your statement could be true, afaik, is if 'finetune' didn't actually offload to GPU when it told me it did (I offloaded 25 out of 29 layers).

File: examples/finetune/finetune.sh (hunk 1 above)
Comment (AndrewGodfrey) [reply examples/finetune/finetune.sh]: I added variables for this. (This was an adventure, ending in filing a [w64devkit issue](https://github.com/skeeto/w64devkit/issues/90)).

**Comment by AndrewGodfrey** [comment]
@ggerganov, What I can say is that it schedules work on the GPU, uses a bunch of VRAM, and when I was developing, it hit not-implemented codepaths (hence the code I added). In the console output it says it's using GPU, and it runs faster (though I don't know if it's doing the same amount of work - it produces a lora that, though working, is different somehow in format than produced by the CPU run).

Are you sure GPU offloading isn't supported? It seems hard to believe that it can do all this AND do the 'real' work on the CPU AND produce a different result AND run faster.

File: examples/finetune/finetune.sh [hunk 2]
```diff
@@ -4,16 +4,15 @@ cd ../..
...
 
 while getopts "dg" opt; do
   case $opt in
     d)
       DEBUGGER="gdb --args"
       ;;
     g)
-      # GPU. The makefile doesn't support CUDA on Windows, so I have to use CMake and so main is built to a different location.
-      # Note: "-d" doesn't really work with this - it will run under gdb, but there are no debugging symbols (in a format gdb understands). I think the easiest workaround is to use WinDbg instead.
```
Comment (kotee4ko) [inline examples/finetune/finetune.sh]: It is possible to add debug generation flags, like:
-ggdb3 -Og -gdwarf-5.

For more legacy gdb versions, even running under windows, we could reduce to -gdwarf-4

**Review by kotee4ko — APPROVED** [review]
Nice work.

File: examples/finetune/finetune.sh (hunk 2 above)
Comment (AndrewGodfrey) [reply examples/finetune/finetune.sh]: It looks like you're talking about gnu compiler flags. Your comments may be relevant to the 'LLAMA_DEBUG' section of the Makefile, which currently uses "-O0 -g".

The problem with CUDA is that we're using CMake, and MSVC compilers, which afaik are unable to produce DWARF info.

**Comment by ggerganov** [comment]
Ah, interesting. I haven't realized that we load the base model with the standard `llama.h` API which can offload some or all the tensors to the GPU. Later when doing finetuning, some of the ops can benefit from the base tensors being already on the GPU although the results from these ops would likely be copied back to the CPU. So this is far from full offloading, but it actually can provide improvement in the speed.

**Review by ggerganov — APPROVED** [review]
Use spaces instead of tabs and we can merge

**Comment by AndrewGodfrey** [comment]
I don't know why it still thinks there are conflicts in ggml.c - I merged from master. And when I click 'Resolve conflicts' I get a weird information-devoid view of ggml.c (which looks the same as before I merged from master).

**Comment by cebtenzzre** [comment]
> I don't know why it still thinks there are conflicts in ggml.c - I merged from master.

It looks like you merged master from your repo and not from this one.

**Review by ggerganov — APPROVED** [review]

**Pull request merged**